onnxruntime/core/optimizer/fast_gelu_fusion.cc

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/fast_gelu_fusion.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph_utils.h"
+#include "float.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+// FastGelu supports limited data types.
+static std::vector<std::string> supported_data_types{"tensor(float16)", "tensor(float)"};
+
+static bool CheckNode(const Node& node, const std::string& op_name, int32_t opset_version, ProviderType provider,
+  bool require_single_output=false){
+  return graph_utils::IsSupportedOptypeVersionAndDomain(node, op_name, {opset_version}) &&
+        node.GetExecutionProviderType() == provider &&
+        optimizer_utils::IsSupportedDataType(node, supported_data_types) &&
+        (!require_single_output || node.GetOutputEdgesCount() == 1);
+}
+
+MatchResult FastGeluFusion::CheckFirstFormula(Graph& graph, Node& mul1_node, 
+  std::vector<std::reference_wrapper<Node>>& nodes_to_fuse) const {
+  MatchResult matchResult{false, nullptr, nullptr, nullptr, nullptr};
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(mul1_node, "Mul", {7}) ||
+      !graph_utils::IsSupportedProvider(mul1_node, GetCompatibleExecutionProviders()) ||
+      mul1_node.GetOutputEdgesCount() != 1 ||
+      !optimizer_utils::IsSupportedDataType(mul1_node, supported_data_types)) {
+    return matchResult;
+  }
+
+  int32_t input_index = -1;
+  const float mul_val = 0.044715f;
+  for (auto i = 0; i < 2; i++) {
+    if (optimizer_utils::IsInitializerWithExpectedValue(graph, *(mul1_node.InputDefs()[i]), mul_val, true)){
+      input_index = i;
+      break;
+    }
+  }
+
+  if (input_index == -1) return matchResult;
+
+  NodeArg* gelu_without_bias_input_arg = mul1_node.MutableInputDefs()[(input_index + 1) % 2];
+  auto p_mul1_input_node = graph_utils::GetInputNode(mul1_node, (input_index + 1) % 2);
+  NodeArg* gelu_input_arg = gelu_without_bias_input_arg;
+  NodeArg* bias_arg = nullptr;
+  if (p_mul1_input_node != nullptr) {
+    Node& mul1_input_node = const_cast<Node&>(*p_mul1_input_node);
+    if (CheckNode(mul1_input_node, "Add", 7, mul1_node.GetExecutionProviderType(), false) &&
+      mul1_input_node.GetOutputEdgesCount() == 4) {
+      gelu_input_arg = mul1_input_node.MutableInputDefs()[0];
+      bias_arg = mul1_input_node.MutableInputDefs()[1];
```

Comment:
Bias need some constraints in input shape, which is missing here.

I would suggest to remove bias related logic in this PR. For bias, it is better to modify bias_gelu_fusion.cc to support FastGelu as well. Otherwise, you might need implement similar input shape checking logic here.

Reply:
Thanks for the comment. I moved the bias part into bias_gelu_fusion.cc. Please review.